### PR TITLE
convert disk, project, and organization lookups to new lookup API

### PR DESCRIPTION
**Pull request by davepacheco**
See #798.  This change:

- addresses all of the remaining resources in the existing Organization/Project/etc. hierarchy.
- updates the Organization datastore interfaces to be consistent with other resources.  (They now take an `authz::Organization` -- expecting the caller to have done the lookup -- rather than a `name`.)
- updates disk_refetch() and instance_refetch() to use the new lookup API, rather than re-implementing their own by-id lookup

There are other resources not covered here but they're all harder to deal with.  See #845.

### CODE REVIEW

File: nexus/src/db/datastore.rs [hunk 1]
```diff
@@ -562,134 +562,26 @@ impl DataStore {
...
-    ///   particular name.  It's just that we don't know at this point whether
-    ///   they're allowed to see the Organization.  It depends on whether, as we
-    ///   look up things inside the Organization, we find something that they
-    ///   _are_ able to see.
-    ///
-    /// **This is an internal-only function.** This function cannot know what
-    /// authz checks are required.  As a result, it should not be made
-    /// accessible outside the DataStore.  It should always be wrapped by
-    /// something that does the appropriate authz check.
-    // TODO-security We should refactor things so that it's harder to
```
Comment (davepacheco) [inline nexus/src/db/datastore.rs:591]: \o/ We finally did this!

File: nexus/src/nexus.rs [hunk 2]
```diff
@@ -745,16 +766,12 @@ impl Nexus {
...
-        let authz_project = self
-            .db_datastore
-            .project_lookup_by_path(organization_name, project_name)
+        let (.., authz_project) = LookupPath::new(opctx, &self.db_datastore)
+            .organization_name(organization_name)
+            .project_name(project_name)
+            .lookup_for(authz::Action::CreateChild)
             .await?;
 
-        // TODO-security This may need to be revisited once we implement authz
```
Comment (davepacheco) [inline nexus/src/nexus.rs:753]: This TODO is no longer applicable.  We did implement authz for saga actions.  The pattern is: we do the authz check before creating the saga for a better user experience, and we also serialize the current authn context, rehydrate it inside the saga, and use that to check again when we run the saga action.  So this TODO is resolved.

I suspect we're not actually authorizing the operations that this saga is doing, but that'll be caught when I sweep through DataStore looking for unprotected functions (which is tracked by #419).

The authorize check isn't removed -- it got moved into the lookup now.

File: nexus/src/nexus.rs [hunk 3]
```diff
@@ -809,19 +824,13 @@ impl Nexus {
...
-            .db_datastore
-            .disk_lookup_by_path(organization_name, project_name, disk_name)
+        let (.., authz_disk) = LookupPath::new(opctx, &self.db_datastore)
+            .organization_name(organization_name)
+            .project_name(project_name)
+            .disk_name(disk_name)
+            .lookup_for(authz::Action::Delete)
             .await?;
 
-        // TODO: We need to sort out the authorization checks.
```
Comment (davepacheco) [inline nexus/src/nexus.rs:817]: Same as above.

**Review by david-crespo — APPROVED** [review]
looks great!

**Comment by davepacheco** [comment]
Thanks for the review!

**Comment by davepacheco** [comment]
GitHub actions looks stuck on the Mac builds after 10+ hours:
Screen Shot 2022-04-01 at 8 44 10 AM

I will try to push a noop commit to try again.

**Comment by davepacheco** [comment]
I'm going to close and open a new PR to try to kick the GitHub Actions.

**Pull request closed without merging**